crates/project_model/src/build_scripts.rs

```diff
@@ -58,7 +58,18 @@ impl WorkspaceBuildScripts {
             cmd.env("RA_RUSTC_WRAPPER", "1");
         }
         cmd.current_dir(workspace.workspace_root());
-        cmd.args(&["check", "--quiet", "--workspace", "--message-format=json"]);
+        cmd.args(&["check", "--quiet", "--message-format=json"]);
+
+        if config.build_script_packages.is_empty() {
+            cmd.arg("--workspace");
+        } else {
+            for pkg in &config.build_script_packages {
+                cmd.arg("-p");
+                cmd.arg(pkg);
+            }
+        }
+
+        cmd.env("IS_RA_BUILDSCRIPT_CHECK", "1");
```

Comment:
I don't think this will work in general. Cargo uses the same cache key for the build invocation with this env var unset and with it set so running `cargo check` to present error messages (flycheck) will likely remove the build script outputs this invocation created if the build script reads this env var. In any case if a build script reads it, every time this invocation or the flycheck invocation runs after running the other will cause the build script to run again and all dependent crates to be recompiled.

Reply:
Yeah this is true. Maybe we should just workaround this in rustc.

Reply:
The buildscripts check usually runs with `RUSTC_WRAPPER=rust-analyzer`; I assume cargo will use a different cache key for that, so it might work? In general, I'm not a fan of build scripts special-casing RA though. And I guess they could just check for `RUSTC_WRAPPER` if they really want to.